client/jetpack-cloud/sections/partner-portal/primary/payment-method-list/index.tsx

```diff
@@ -13,20 +15,65 @@ import StoredCreditCardLoading from 'calypso/jetpack-cloud/sections/partner-port
 import {
 	getAllStoredCards,
 	isFetchingStoredCards,
+	hasMoreStoredCards,
 } from 'calypso/state/partner-portal/stored-cards/selectors';
+import type { ReactElement } from 'react';
+
 import './style.scss';
 
+const ITEMS_PER_PAGE = 8;
+
+const preparePagingCursor = (
+	direction: 'next' | 'prev',
+	items: PaymentMethod[],
+	isFirstPage: boolean
+) => {
+	if ( ! items.length || isFirstPage ) {
+		return {
+			startingAfter: '',
+			endingBefore: '',
+		};
+	}
+
+	return {
+		startingAfter: direction === 'next' ? items[ items.length - 1 ].id : '',
+		endingBefore: direction === 'prev' ? items[ 0 ].id : '',
+	};
+};
+
 export default function PaymentMethodList(): ReactElement {
 	const translate = useTranslate();
 	const storedCards = useSelector( getAllStoredCards );
 	const isFetching = useSelector( isFetchingStoredCards );
+	const hasMoreItems = useSelector( hasMoreStoredCards );
 	const cards = storedCards.map( ( card: PaymentMethod ) => (
 		<StoredCreditCard key={ card.id } card={ card } />
 	) );
 
+	const [ hasMoreItemsInitial, setHasMoreItemsInitial ] = useState( false );
+	const [ paging, setPaging ] = useState( { startingAfter: '', endingBefore: '' } );
+	const [ currentPage, setCurrentPage ] = useState( 1 );
+	const onPageClick = ( pageNumber: number ) => {
+		const direction = pageNumber > currentPage ? 'next' : 'prev';
+
+		// Set a cursor for use in pagination.
+		setPaging( preparePagingCursor( direction, storedCards, pageNumber === 1 ) );
+
+		setCurrentPage( pageNumber );
+	};
+
+	// If the initial request has more items, we need to always display the pagination controls.
+	useEffect( () => {
+		if ( hasMoreItems ) {
+			setHasMoreItemsInitial( true );
+		}
+	}, [ hasMoreItems ] );
```

Comment:
This seems unnecessary - we can use  useState( hasMoreItems ) as useState will only respect the passed value the first time it's called so even if hasMoreItems changes over time it won't matter to the useState call:
```
const [ showPagination ] = useState( hasMoreItems );
```